compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileAndDistribute.cpp

```diff
@@ -249,31 +246,86 @@ static void populatePromotionPatterns(MLIRContext *context,
           .addFilter(filterFunction));
 }
 
+static bool propagateIntoProducer(memref::CopyOp copyOp) {
+  // Look for a fill Op writing into the copyOp source.
+  Operation *prevOp = copyOp->getPrevNode();
+  while (prevOp) {
+    if (isSideEffectFree(prevOp)) {
+      prevOp = prevOp->getPrevNode();
+      continue;
+    }
+
+    auto fillOp = dyn_cast<linalg::FillOp>(prevOp);
+    if (!fillOp) break;
+    if (fillOp.output() != copyOp.getSource()) break;
+    // Move the fillOp and change the destination to the copy destination.
+    fillOp->moveBefore(copyOp);
+    fillOp.getOutputsMutable().assign(copyOp.getTarget());
+    return true;
+  }
+  return false;
+}
+
+// Split input/output operand from copy from shared memory into a separate
+// input.
+static void mergeCopySourceIntoGeneric(Value source, linalg::GenericOp op) {
+  SmallVector<Value> newOperands;
+  SmallVector<AffineMap> maps;
+  for (auto in : op.getInputOperands()) {
+    newOperands.push_back(in->get());
+    maps.push_back(op.getTiedIndexingMap(in));
+  }
+  newOperands.push_back(source);
+  maps.push_back(op.getTiedIndexingMap(*op.getOutputOperands().begin()));
+  maps.push_back(op.getTiedIndexingMap(*op.getOutputOperands().begin()));
```

Comment:
This effectively assumes the generic op has identity map (given we use the same map for the new added `value`, which is coming form copy, and the original output)? I didn't see such checks in the `propagateIntoConsumer` function.. We should probably directly use identify map for `value` here?

Reply:
I don't think it does, the operand was input/output and both were using the same indexing map, now that I separate out the input into its own operand it should have the same indexing map.
Since I currently don't have a test for that I added a restriction into `propagateIntoConsumer` but I think we can keep the code generic here.